### PR TITLE
Simplify and fix hack/{verify,update}-staging-{client-go,godeps}.sh

### PR DESCRIPTION
**Pull request by sttts**
- merge `hack/{verify,update}-staging-client-go.sh`
- pin godep with shared code
- remove godep-restore completely from the process and replace with a simple check that godeps are restored
- add safety check in `staging/copy.sh` that there is no lingering `k8s.io/apimachinery` in the GOPATH which would lead to inconsistent client-go builds (!)
- check that all these scripts only operate in a clean working dir.

### CODE REVIEW

File: staging/copy.sh [hunk 1]
```diff
@@ -20,22 +20,30 @@ set -o pipefail
...
       echo "Invalid option: -$OPTARG" >&2
+      exit 1
       ;;
   esac
 done
 readonly FAIL_ON_CHANGES DRY_RUN
 
-echo "**PLEASE** run \"godep restore\" before running this script"
+if [ "${RUN_FROM_UPDATE_SCRIPT}" != true ]; then
+  echo "WARNING WARNING WARNING - do not run this script directly, but via hack/update-staging-client-go.sh"
```
Comment (deads2k) [inline staging/copy.sh]: no exit?

File: hack/update-staging-client-go.sh [hunk 2]
```diff
@@ -18,5 +18,52 @@ set -o errexit
...
+    GODEP_RESTORE_GOPATH="${GODEP_RESTORE_TMPDIR}:${GOPATH}"
+else
+    GODEP_RESTORE_GOPATH="${GOPATH}"
+fi
+
+echo "Running godep restore"
+GOPATH="${GODEP_RESTORE_GOPATH}" godep restore ${V} 2>&1 | sed 's/^/  /'
+
+echo "Running staging/copy.sh"
+eval staging/copy.sh ${COPY_FLAGS} 2>&1 | sed 's/^/  /'
```
Comment (sttts) [inline hack/update-staging-client-go.sh]: this is wrong. We have to set the GOPATH

File: hack/update-staging-client-go.sh [hunk 3]
```diff
@@ -18,5 +18,52 @@ set -o errexit
 set -o nounset
 set -o pipefail
 
+V=""
+IN_PLACE="0"
```
Comment (deads2k) [inline hack/update-staging-client-go.sh]: What is in place?

File: hack/update-staging-client-go.sh [hunk 4]
```diff
@@ -18,5 +18,52 @@ set -o errexit
...
+    }
+    trap cleanup EXIT SIGINT
+    mkdir -p "${GODEP_RESTORE_TMPDIR}/src"
+    GODEP_RESTORE_GOPATH="${GODEP_RESTORE_TMPDIR}:${GOPATH}"
+else
+    GODEP_RESTORE_GOPATH="${GOPATH}"
+fi
+
+echo "Running godep restore"
+GOPATH="${GODEP_RESTORE_GOPATH}" godep restore ${V} 2>&1 | sed 's/^/  /'
```
Comment (deads2k) [inline hack/update-staging-client-go.sh]: what does the sed do?

File: hack/update-staging-client-go.sh [hunk 5]
```diff
@@ -18,5 +18,52 @@ set -o errexit
...
+      V="-v"
+      ;;
+    i)
+      IN_PLACE=1
+      ;;
+    d)
+	  COPY_FLAGS+=" -d"
+	  ;;
+    f)
+	  COPY_FLAGS+=" -f"
```
Comment (deads2k) [inline hack/update-staging-client-go.sh]: what is -f?

File: staging/copy.sh (hunk 1 above)
Comment (sttts) [inline staging/copy.sh]: Agree.

File: hack/update-staging-client-go.sh (hunk 3 above)
Comment (sttts) [inline hack/update-staging-client-go.sh]: = no temp dir for godep-restore

File: hack/update-staging-client-go.sh (hunk 4 above)
Comment (sttts) [inline hack/update-staging-client-go.sh]: indent

File: hack/update-staging-client-go.sh (hunk 5 above)
Comment (sttts) [inline hack/update-staging-client-go.sh]: passed through to copy.sh: fail on diff.

**Comment by sttts** [comment]
@eparis can you take a look with your bash-wizard eyes?

File: hack/lib/util.sh [hunk 6]
```diff
@@ -411,6 +411,71 @@ kube::util::git_upstream_remote_name() {
...
+# Ensure that the given godep version is installed and in the path
+kube::util::ensure_godep_version() {
+  if [[ "$(godep version)" == *"godep ${1}"* ]]; then
+    return
+  fi
+
+  kube::util::ensure-temp-dir
+  mkdir -p "${KUBE_TEMP}/go/src"
+  GOPATH="${KUBE_TEMP}/go" go get -u github.com/tools/godep 2>/dev/null
+  (cd "${KUBE_TEMP}/go/src/github.com/tools/godep" && git checkout "${1:-v74}" && GOPATH="${KUBE_TEMP}/go" go install .)
```
Comment (eparis) [inline hack/lib/util.sh]: would use be willing to use pushd/popd here, to make this more readable?

File: hack/lib/util.sh [hunk 7]
```diff
@@ -411,6 +411,71 @@ kube::util::git_upstream_remote_name() {
...
+  PATH="${KUBE_TEMP}/go/bin:${PATH}"
+  hash -r # force bash to clear PATH cache
```
Comment (eparis) [inline hack/lib/util.sh]: Isn't just updating PATH enough?

File: hack/update-staging-godeps.sh [hunk 8]
```diff
@@ -23,102 +23,80 @@ set -o errexit
...
-	echo "rewriting Godeps.json to remove commits that don't really exist because we haven't pushed the prereqs yet"
-	GOPATH="${ORIGINAL_GOPATH}"
-	go run "${KUBE_ROOT}/staging/godeps-json-updater.go" --godeps-file="${TARGET_DIR}/src/${repo}/Godeps/Godeps.json" --client-go-import-path="${repo}"
-
-	popd
-	echo "finished ${repo} save"
+  local repo="${1}"
+
+  echo "Updating godeps for ${repo}"
+  rm -rf Godeps # remove the current Godeps.json so we always rebuild it
```
Comment (eparis) [inline hack/update-staging-godeps.sh]: I don't love that this function assumed PWD == ${repo}.  should we pushd/popd just to be safe against future users? its a minor nit for sure.

File: hack/update-staging-godeps.sh [hunk 9]
```diff
@@ -23,102 +23,80 @@ set -o errexit
...
 
+V=""
+DRY_RUN=false
+FAIL_ON_DIFF=false
+while getopts ":vdf" opt; do
+  case $opt in
+    v) # increase verbosity
+      V="-v"
+      ;;
+    i) # do not godep-restore into a temporary directory, but use the existing GOPATH
```
Comment (eparis) [inline hack/update-staging-godeps.sh]: Why is this `i)`?

File: hack/verify-staging-client-go.sh [hunk 10]
```diff
@@ -18,65 +18,12 @@ set -o errexit
...
-export GOPATH="${TEMP_STAGING_GOPATH}"
-godep restore ${V} 2>&1 | sed 's/^/  /'
+echo "Smoke testing client-go examples"
+go install ./staging/src/k8s.io/client-go/examples/... 2>&1 | sed 's/^/  /'
 
-echo "Testing staging/copy.sh"
-staging/copy.sh -d 2>&1 | sed 's/^/  /'
-popd
+# Run update-staging-client.sh in dry-run mode, copy nothing into the staging dir
+hack/update-staging-client-go.sh "$@" -d
```
Comment (eparis) [inline hack/verify-staging-client-go.sh]: For consistency can we put `-d` before the `$@`?

File: hack/lib/util.sh (hunk 6 above)
Comment (sttts) [inline hack/lib/util.sh]: sure

File: hack/lib/util.sh (hunk 7 above)
Comment (sttts) [inline hack/lib/util.sh]: I thought so too. Looks like latest bash has PATH caching. Took me a while to understand why it was always launching godep v79 instead of v74.

File: hack/update-staging-godeps.sh (hunk 9 above)
Comment (sttts) [inline hack/update-staging-godeps.sh]: good catch. Left over from old code. Removing.

File: hack/update-staging-godeps.sh (hunk 8 above)
Comment (sttts) [inline hack/update-staging-godeps.sh]: Makes sense. I think we have that in a number of scripts.

File: hack/verify-staging-client-go.sh (hunk 10 above)
Comment (sttts) [inline hack/verify-staging-client-go.sh]: 👍

File: hack/update-staging-godeps.sh (hunk 9 above)
Comment (sttts) [inline hack/update-staging-godeps.sh]: in fact, this is just a typo: `d)`

**Comment by sttts** [comment]
@eparis addressed your comments

**Comment by sttts** [comment]
Fixed a typo, forgotten `-u` when calling copy.sh.

**Comment by eparis** [comment]
/lgtm
just to be safe

File: hack/update-staging-godeps.sh [hunk 11]
```diff
@@ -23,102 +23,84 @@ set -o errexit
...
-	GOPATH="${ORIGINAL_GOPATH}"
-	go run "${KUBE_ROOT}/staging/godeps-json-updater.go" --godeps-file="${TARGET_DIR}/src/${repo}/Godeps/Godeps.json" --client-go-import-path="${repo}"
-
-	popd
-	echo "finished ${repo} save"
+  local repo="${1}"
+  pushd "${TMP_GOPATH}/src/k8s.io/${repo}" >/dev/null
+    echo "Updating godeps for k8s.io/${repo}"
+    rm -rf Godeps # remove the current Godeps.json so we always rebuild it
+    GOPATH="${TMP_GOPATH}:${GOPATH}:${KUBE_ROOT}/staging" godep save ${V} ./... 2>&1 | sed 's/^/  /'
```
Comment (caesarxuchao) [inline hack/update-staging-godeps.sh]: If I have a different copy of apimachinery in ${GOPATH}, will godep save prefer that copy over the one in ${KUBE_ROOT}/staging?

File: hack/verify-staging-client-go.sh [hunk 12]
```diff
@@ -18,65 +18,12 @@ set -o errexit
...
-pushd "${TEMP_STAGING_GOPATH}/src/k8s.io/kubernetes" >/dev/null
-export GOPATH="${TEMP_STAGING_GOPATH}"
-godep restore ${V} 2>&1 | sed 's/^/  /'
+echo "Smoke testing client-go examples"
+go install ./staging/src/k8s.io/client-go/examples/... 2>&1 | sed 's/^/  /'
 
-echo "Testing staging/copy.sh"
-staging/copy.sh -d 2>&1 | sed 's/^/  /'
-popd
+# Run update-staging-client.sh in dry-run mode, copy nothing into the staging dir
```
Comment (caesarxuchao) [inline hack/verify-staging-client-go.sh]: Could you add a comment saying the purpose to verify that `hack/update-staging-client-go.sh` can be run successfully, not that verifying if hack/update-staging-client-go.sh needs to be run?
(ignore me if you have done that in another PR already)

File: hack/update-staging-godeps.sh (hunk 11 above)
Comment (caesarxuchao) [inline hack/update-staging-godeps.sh]: i.e., Do you need something like https://github.com/kubernetes/kubernetes/pull/41987/files#diff-79c7edfbbe6b22642e05ac8c9cd431e7R124 before running `godep save`?

File: hack/lib/util.sh (hunk 7 above)
Comment (caesarxuchao) [inline hack/lib/util.sh]: huh, this got me so many times when i tried to use my freshly built kubectl.

File: hack/update-staging-godeps.sh (hunk 11 above)
Comment (caesarxuchao) [inline hack/update-staging-godeps.sh]: Probably it doesn't hurt to have another copy in GOPATH, because the publish robot will rewrite the revision later. If that's the case, can we remove https://github.com/kubernetes/kubernetes/pull/41987/files#diff-79c7edfbbe6b22642e05ac8c9cd431e7R124?

**Comment by caesarxuchao** [comment]
@sttts sorry i didn't get to this yesterday. Really glad to see we get rid of two `godep restore`! I left some comments on whether it's ok to have a copy of k8s.io/xxx in GOPATH before running `godep save` (i think it's ok).

File: hack/update-staging-godeps.sh (hunk 11 above)
Comment (sttts) [inline hack/update-staging-godeps.sh]: I think you are right. With `${TMP_GOPATH}:${KUBE_ROOT}/staging:${GOPATH}` we should get rid of the test above. Wondering why I didn't do that.

Comment (sttts) [inline hack/update-staging-godeps.sh]: No, it does not work. Godep does not find the git repo if we change the order. So we also need the check for the apimachinery repo in the other script :-/

Comment (sttts) [inline hack/update-staging-godeps.sh]: Adding the check is not so easy, at least not with `go list`. In CI we have a clean GOPATH, so not lingering staging repos. So CI will catch any inconcistency. I propose we leave it as it is now. We can iterate late. @caesarxuchao wdyt?

File: hack/verify-staging-client-go.sh (hunk 12 above)
Comment (sttts) [inline hack/verify-staging-client-go.sh]: In a follow-up PR (after freeze) we will add the `-f` flag to enforce correct client-go. Let's leave it like it is for now. The follow-up will turn it into a real diff-check.

Comment (sttts) [inline hack/verify-staging-client-go.sh]: https://github.com/kubernetes/kubernetes/pull/42053

File: hack/update-staging-godeps.sh (hunk 11 above)
Comment (caesarxuchao) [inline hack/update-staging-godeps.sh]: Sure. Thanks.

**Comment by caesarxuchao** [comment]
Get this error message running `hack/update-staging-client-go.sh` after cherry-picking the commits here, 
```
  hg parents: option -T not recognized
  Unexpected HEAD 'hg parents [-r REV] [FILE]
  
  show the parents of the working directory or revision
  
  options:
  
   -r --rev REV           show parents of the specified revision
      --style STYLE       display using template map file
      --template TEMPLATE display with template
  
  use "hg help parents" to show the full help text' at /usr/local/google/home/xuchao/go-workspace/src/bitbucket.org/ww/goautoneg, expected '75cd24fc2f2c2a2088577d12123ddee5f54e0675'.
```

I'll try to debug it and report back.

**Comment by caesarxuchao** [comment]
`-T` doesn't work for my version of hg, `--template` works

```
$ hg --version
Mercurial Distributed SCM (version 2.8.2)
```

**Comment by sttts** [comment]
@caesarxuchao 
```
Mercurial Distributed SCM (version 4.1)
(see https://mercurial-scm.org for more information)
```

Fixing.

**Comment by sttts** [comment]
@k8s-bot kops aws e2e test this

**Comment by sttts** [comment]
@k8s-bot non-cri e2e test this
